mypy/newsemanal/semanal.py

```diff
@@ -381,12 +403,30 @@ def file_context(self, file_node: MypyFile, fnam: str, options: Options,
         del self.options
 
     def visit_func_def(self, defn: FuncDef) -> None:
+        self.add_func_to_symbol_table(defn)
+
         if not self.recurse_into_functions:
             return
 
         with self.scope.function_scope(defn):
             self._visit_func_def(defn)
 
+    def add_func_to_symbol_table(self, func: FuncDef) -> None:
+        if self.is_func_scope():
+            # These are handled later.
+            return
+        func._fullname = self.qualified_name(func.name())
+        # TODO: Module-level __getattr__
+        # TODO: Conditional function definitions
+        if self.is_module_scope():
+            if func.name() not in self.globals:
+                self.globals[func.name()] = SymbolTableNode(GDEF, func)
+        elif self.is_class_scope():
+            info = self.type
+            assert info is not None
+            if func.name() not in info.names:
+                info.names[func.name()] = SymbolTableNode(MDEF, func)
```

Comment:
I think we are making the same mistake again here. We already have several functions and code fragments for adding symbols to symbol table. I would rather have _one_ method that adds symbol node to current namespace. It can have some additional flags and can be called from all similar places like this.

Reply:
Yes, this will need to be refactored. This is just a placeholder implementation.